tilings/algorithms/minimal_gridded_perms.py

```diff
@@ -0,0 +1,123 @@
+        """Check if a gridded permutation contains all requirements."""
+        return all(gp.contains(*req) for req in self.requirements)
+
+    def yielded_subgridded_perm(self, gp: GriddedPerm) -> bool:
+        """Return True if a subgridded perm was yielded."""
+        return gp.contains(*self.yielded)
+
+    def prepare_queue(self) -> None:
+        """Add cell counters with gridded permutations to the queue."""
+        if len(self.requirements) <= 1:
+            return
+        first_req = self.requirements[0]
+
+        for gp in first_req:
+            max_cell_counts = []
+            cells = set()
+            for gps in product(*self.requirements[1:]):
+                max_cell_count = defaultdict(int)
+                for othergp in gps:
+                    max_cell_count = self.cell_counter(othergp, max_cell_count)
+                max_cell_count = self.cell_counter(gp, max_cell_count)
+                max_cell_counts.append(max_cell_count)
+                cells.update(max_cell_count)
+
+            new_info = Info([gp, max_cell_counts, cells])
+            self.put(new_info)
+
+    @staticmethod
+    def cell_counter(
+                     gp: GriddedPerm, counter: Optional[Counter] = None
+                     ) -> Counter:
+        """Returns a counter of the number of cells used by the gridded
+        permutation. If 'counter' is given, will update this Counter."""
+        if counter is None:
+            counter = defaultdict(int)
+        for cell in gp.pos:
+            counter[cell] += 1
+        return counter
+
+    def get_cells_to_try(
+                         self, gp: GriddedPerm, max_cell_counts: List[Counter],
+                         cells: List[Cell]) -> Iterator[Cell]:
+        """Yield cells that a gridded permutation could be extended by."""
+        currcellcounts = self.cell_counter(gp)
+        if any(all(count >= max_cell_count[cell]
+                   for cell, count in currcellcounts.items())
+               for max_cell_count in max_cell_counts):
+            return
+        for cell in cells:
+            if any(currcellcounts[cell] < max_cell_count[cell]
+                   for max_cell_count in max_cell_counts):
+                yield cell
+
+    def minimal_gridded_perms(self) -> Iterator[GriddedPerm]:
+        """Yield all minimal gridded perms on the tiling."""
+        if not self.requirements:
+            return
+        if len(self.requirements) == 1:
+            yield from iter(self.requirements[0])
+            return
+
+        self.prepare_queue()
+        while not self.empty_queue():
+            startgp, max_cell_counts, cells = self.pop()
+            for cell in self.get_cells_to_try(startgp, max_cell_counts, cells):
+                for gp in startgp.insert_point(cell):
+                    if gp not in self.seen:
+                        self.seen.add(gp)
+                        if (self.satisfies_obstructions(gp) and
```

Comment:
You probably have the answer to this. If a permutation have been yielded then it has been seen. Isn't the yielded check redundant with the seen test?

Reply:
Potentially a yielded gridded permutation could be contained in the new gridded perm. As we are not building left to right, but jumping around cells, I can't guarantee there is a unique path to each gridded permutation we come to.

Reply:
Ah I did not realize the **sub** part